### PR TITLE
arch: arc: enable the workaround of sleep only for SMP case in nsim

### LINKED ISSUE #24276: tests/kernel/context hangs on ARC in test_kernel_cpu_idle
In all cases the test times out with no output.
Observed on nsim_sem, nsim_em, nsim_sem_mpu_stack_guard, nsim_em_mpu_stack_guard

```
===================================================================
starting test - test_kernel_cpu_idle
```

### PR DESCRIPTION
**Pull request by vonhust**
Because the issue of nsim, the sleep instruction doest not work correctly
when SMP is enabled. A workaround is introduced in commit d56a12d955,
this workaround should be enabled only for SMP case in nsim.

For other cases, no need of this workaround.

This commit fixes #24276

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

**Review by abrodkin — CHANGES_REQUESTED** [review]
Can we invert that conditional code selection so that it becomes easier for reading and understanding?
I.e. something like that:
```c
#if defined(CONFIG_SOC_NSIM) && defined(CONFIG_SMP)
	seti r1
#else
	sleep r1
#endif
```

From the code above a reader may easily conclude that `seti r1` only makes sense for SMP nSIM, for all the rest we use `sleep r1`.

File: tests/kernel/context/testcase.yaml [hunk 1]
```diff
@@ -4,4 +4,4 @@ tests:
     min_ram: 20
     # FIXME: Remove `qemu_cortex_r5` exclusion once Zephyr SDK 0.11.3 is
     #        mainlined (see #22904).
-    platform_exclude: qemu_cortex_r5
+    platform_exclude: qemu_cortex_r5 nsim_hs_smp
```
Comment (andyross) [inline tests/kernel/context/testcase.yaml:7]: This is a big test to exclude, tests/kernel/context really needs to be passing.  If you absolutely need a workaround, can't you do a more targetted whitelisting, maybe with a ztest_test_skip() at the specific test case?

**Review by andyross — CHANGES_REQUESTED** [review]

**Review by abrodkin — CHANGES_REQUESTED** [review]
It's not really clear why https://github.com/zephyrproject-rtos/zephyr/pull/24302/commits/c6e2c595d7ce73b51c319c276c48547654e1b0f9 is needed and how problem in `sleep` simulation (being worked-arounded in https://github.com/zephyrproject-rtos/zephyr/pull/24302/commits/3c2d1801a7a044031824a17a8d9eb0a022bb95a7) leads to black-listing `tests/kernel/context` for nSIM with SMP. May we have more meaningful explanation for that here?

File: tests/kernel/context/testcase.yaml (hunk 1 above)
Comment (vonhust) [reply tests/kernel/context/testcase.yaml:7]: make senses, i will change

**Comment by vonhust** [comment]
> Can we invert that conditional code selection so that it becomes easier for reading and understanding?
> I.e. something like that:
> 
> ```c
> #if defined(CONFIG_SOC_NSIM) && defined(CONFIG_SMP)
> 	seti r1
> #else
> 	sleep r1
> #endif
> ```
> 
> From the code above a reader may easily conclude that `seti r1` only makes sense for SMP nSIM, for all the rest we use `sleep r1`.

Make senses, already changed

**Comment by vonhust** [comment]
> It's not really clear why [c6e2c59](https://github.com/zephyrproject-rtos/zephyr/commit/c6e2c595d7ce73b51c319c276c48547654e1b0f9) is needed and how problem in `sleep` simulation (being worked-arounded in [3c2d180](https://github.com/zephyrproject-rtos/zephyr/commit/3c2d1801a7a044031824a17a8d9eb0a022bb95a7)) leads to black-listing `tests/kernel/context` for nSIM with SMP. May we have more meaningful explanation for that here?

I will add more explanation in commit msg

**Comment by andrewboie** [comment]
I'm confused too. The change in cpu_idle.S rearranges some ifdefs for clarity, but no functional changes. Need clarification on why the test_kernel_cpu_idle case needs to be skipped.

**Review by tyrantlizard77 — DISMISSED** [review]
[edit] posted in wrong PR

**Review by ruuddw — APPROVED** [review]
Still a workaround, but a clean one now. Is there an issue to track the nsim problem, so we don't forget fixing it?

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
@andyross @abrodkin can you re-visit your reviews?

**Review by abrodkin — CHANGES_REQUESTED** [review]
Similarly to @ruuddw I would strongly suggest to mention a reference to internal bug report (Jira reference) in the code so that we may get back to it once it's solved or rather ping nSIM people every time we bump into it :)

**Comment by ruuddw** [comment]
Any reason why this one hasn't been merged yet?

**Comment by abrodkin** [comment]
@ruuddw I asked for a reference to internal bug report against nSIM, also there seem to be something requested by @andyross.

**Review by andyross — APPROVED** [review]
My -1 was stale, sorry.  I was complaining that we shouldn't be whitelisting tests/kernel/context (frankly one of the most important tests in the suite for an arch layer), and that was fixed long ago.

**Comment by carlescufi** [comment]
@vonhust and @abrodkin: @andyross has approved, could you add the reference you lack so we can merge this?

**Comment by vonhust** [comment]
I add snps jira issue P10019563-41294 to trace the bug in nsim.

**Comment by carlescufi** [comment]
@abrodkin since @vonhust added the reference to your internal Jira, can you approve?

**Comment by abrodkin** [comment]
@carlescufi @vonhust I meant reference in comments in the code so that it is not lost and we know where to look at when examining the code next time. But there seems to be no changes in the code itself.

**Comment by vonhust** [comment]
@abrodkin  already add a record in code comments of cpu_idle.s

**Review by abrodkin — APPROVED** [review]
LGTM now!

**Pull request merged**